### ISSUE TITLE
mirage 4.0 modifies my global opam switch without permission

### ISSUE THREAD

**Issue by kit-ty-kate** [mirage/mirage#1284]
I’m trying to port a mirage 3.0 project to test mirage 4.0.
So far:
```
$ mirage configure
```
worked fine. But:
```
$ make deps
```
(after battling around before realising that for some reason it doesn’t like OCaml 4.14), is modifing my global 4.13 switch and at no point asked me if it was ok to install/upgrade/reinstall every packages in my switch. I think this should not be the case.

**Comment by dinosaure** [comment]
Yeah, the current `make depends` tries to install `ocaml-freestanding` with `opam install -y`, so we don't ask the user if he/she wants that. Not sure that it's a good design or not but it seems clear for us (as @TheLortex said) that the problem is more about the optional dependency of `ocaml-freestanding` into several third-part packages.

Currently, such dependency is not needed anymore (see #1262) and we should lint these packages and delete this final dependency. The result will be a healthy eco-system where nobody should depend on `ocaml-freestanding`. By this way, the `opam install -y ocaml-freestanding` should just add `ocaml-freestanding` and should not recompile all of your libraries.

**Comment by hannesm** [comment]
Now that I think a bit more about it, maybe it would be nicer to rename ocaml-freestanding (0.7.0)?

Since the situation with MirageOS3 is that various packages have an optional dependency on it (mirage-crypto-ec, ...). They use the `pkg-config` file from ocaml-freestanding (< 0.7.0) for cross-compilation.

Now, with the MirageOS4 approach, the cross-compilation is done by dune. The ocaml-freestanding (0.7.0+) no longer provide any pkg-config files. This means the "correct" dependency would be:

`depopt: ocaml-freestanding < 0.7.0`

But I think this is tricky to express in opam, unfortunately. Any more recent ocaml-freestanding can happily be installed with mirage-crypto-ec (i.e. there's no conflict for ocaml-freestanding >= 0.7.0).

TL;DR: if we remove ocaml-freestanding 0.7.0 from opam-repository and release it as mirage-ocaml (or mirage-freestanding-ocaml or ocaml-freestanding4) [and include a conflict with ocaml-freestanding in that package], that would solve the above issues, and no recompilation would be done.

WDYT?

**Comment by kit-ty-kate** [comment]
Why is it a depopt in the first place if it’s not used?

**Comment by hannesm** [comment]
@kit-ty-kate in MirageOS3, ocaml-freestanding is used for cross-compilation (outputting a `libmirage_crypto_ec_freestanding_stubs.a` that is linked into the unikernel instead of the `libmirage_crypto_ec_stubs.a` (compiled for the host platform)).

With MirageOS4, all of that changes (and is done by dune in a duniverse folder) -- i.e. at some point we can also remove the MirageOS3 cross-compilation Makefiles (and remove the depopts). But for the time being, some of these packages will be compatible with both MirageOS3 and MirageOS4.

Anyways, there are likely other people who're deeper in that MirageOS3 -> MirageOS4 work who have better ideas what to do about your issue "the whole opam switch is recompiling".

**Comment by hannesm** [comment]
I guess #1294 will solve this issue.
